### PR TITLE
Add an emitter package

### PR DESCRIPTION
**Pull request by scruffian**
#### Changes proposed in this Pull Request

* Emitter is reused in other libs that I want to extract (PopupMonitor), so I'm extracting this first

#### Testing instructions

* Running npm install to create the package
* Then run npm start to boot Calypso
* Try making a new connection at /marketing/connections - this relies on PopupMonitor which in turn relies on the Emitter package.

### CODE REVIEW

**Comment by blowery** [comment]
There should be a shrinkwrap change here too. Did it change after you ran `npm install`?

**Comment by scruffian** [comment]
> There should be a shrinkwrap change here too. Did it change after you ran npm install?

added, thanks

File: packages/emitter/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Emitter
+
+This utility function allows us to listen for changes to a store and then fire events when the store changes.
+
+### Usage
+
+```js
+import { emitter } from '@automattic/emitter';
+const Store = createAStore();
+emitter( Store );
```
Comment (sirreal) [inline packages/emitter/README.md]: How do folks feel about this API where we mutate the object to add `EventEmitter` methods? `emitter( obj )` returns undefined, which may make it tricky to reason about (and makes it very difficult to type 🙂).

**Review by sirreal — COMMENTED** [review]
On one hand, this is simply moving some code around, which is nice.

On the other hand, this is the entire implementation:

https://github.com/Automattic/wp-calypso/blob/83d93e9f44f93a9425df3d0974af8679102b3f42/packages/emitter/src/index.js#L10-L16

There seem to be several places that could be improved:

- `emitChange` should likely be `return this.emit( 'change' )` instead of returning `undefined`.
- `Object.assign( prototype, EventEmitter.prototype )` seems pretty strange. Keep in mind that the first argument is an object (not its prototype) and the second _is_ a prototype…
- We add `off`, which [seems to already exist on `EventEmitter`](https://github.com/Gozala/events/blob/7b911040bd1c5927c8b68ab86eda14a467931f26/events.js#L322).

A general note, I'd like to see this implemented in TypeScript and I'd love to have the type system know what EventEmitter properties are available on the emitter-decorated objects 🙂 

---

I'm fine with seeing this move ahead as a v1, but I wanted to note this in case folks are interested in making some improvements. 2 of my points seem like low-hanging fruit.

**Review by sgomes — CHANGES_REQUESTED** [review]
I'm reluctant about consolidating the mixin approach here into a package, rather than refactoring the code to not use it. This approach has long been deprecated in React, as it has a number of issues. See [Mixins considered harmful](https://reactjs.org/blog/2016/07/13/mixins-considered-harmful.html).

I understand the need for legacy code to exist while it doesn't get refactored, but this can be seen as going in the other direction, and may actually be perceived as a legitimisation of the mixin approach.

There isn't a huge number of components in need of a change here, so one option is to refactor every component making use of this mixin to use composition instead. That could be done with a new library/package that could provide a factory that gets used to create an event emitter instance, which gets saved in the component instance, for example. I'm happy to participate in any refactoring efforts.

Another option to consider is simply copying and pasting the code (while refactoring it, maybe) for the component you're hoping to extract, instead of refactoring everything. It may be out of scope for you to perform the large-scale refactor, which I completely understand, so that's perfectly fine. I just don't think we should create a new package with a deprecated approach, i.e. mixins 😕

**Comment by jsnajdr** [comment]
The only job the emitter package does is to make a class inherited from `EventEmitter`. Doing that is not entirely trivial when working directly with prototypes, so there are helper packages for that.

We use `inherit` in Calypso, which can be used to completely replace this. Also, ES6 class syntax makes the job trivial:
```js
class Foo extends EventEmitter {}
```
In this case, the JS runtime or Babel does all the prototype setup under the hood.

I'm in favor of removing `lib/mixins/emitter` completely and replacing it with one of the alternatives.

The `emitChange()` calls will need to be replaced with `emit( 'change' )`. It looks like an easy search and replace task.

It's not even a React mixin, as the name might suggest 🙂

**Comment by sgomes** [comment]
> It's not even a React mixin, as the name might suggest 🙂

It's still the mixin approach, however, as it modifies the prototype of the receiving class to add new methods. The pitfalls of the mixin approach still apply 😕

**Pull request closed without merging**